kernel: include/linux/list.h:1063:2: error: attribute 'error' is already applied with different arguments

When using clang-extract to extract a function from ext4 on current upstream kernel, using the commands below:

```sh
/usr/local/bin/clang-extract -Wp,-MMD,fs/ext4/.super.o.d -nostdinc -I./arch/x86/include -I./arch/x86/include/generated -I./include -I./arch/x86/include/uapi -I./arch/x86/include/generated/uapi -I./include/uapi -I./include/generated/uapi -include ./include/linux/compiler-version.h -include ./include/linux/kconfig.h -include ./include/linux/compiler_types.h -D__KERNEL__ -fmacro-prefix-map=./= -Werror -std=gnu11 -fshort-wchar -funsigned-char -fno-common -fno-PIE -fno-strict-aliasing -mno-sse -mno-mmx -mno-sse2 -mno-3dnow -mno-avx -fcf-protection=branch -fno-jump-tables -m64 -falign-jumps=1 -falign-loops=1 -mno-80387 -mno-fp-ret-in-387 -mpreferred-stack-boundary=3 -mskip-rax-setup -mtune=generic -mno-red-zone -mcmodel=kernel -Wno-sign-compare -fno-asynchronous-unwind-tables -mindirect-branch=thunk-extern -mindirect-branch-register -mindirect-branch-cs-prefix -mfunction-return=thunk-extern -fno-jump-tables -fpatchable-function-entry=16,16 -fno-delete-null-pointer-checks -O2 -fno-allow-store-data-races -fstack-protector-strong -ftrivial-auto-var-init=zero -fno-stack-clash-protection -pg -mrecord-mcount -mfentry -DCC_USING_FENTRY -falign-functions=16 -fstrict-flex-arrays=3 -fno-strict-overflow -fno-stack-check -fconserve-stack -Wall -Wundef -Werror=implicit-function-declaration -Werror=implicit-int -Werror=return-type -Werror=strict-prototypes -Wno-format-security -Wno-trigraphs -Wno-frame-address -Wno-address-of-packed-member -Wmissing-declarations -Wmissing-prototypes -Wframe-larger-than=2048 -Wno-main -Wno-unused-but-set-variable -Wno-unused-const-variable -Wno-dangling-pointer -Wvla -Wno-pointer-sign -Wcast-function-type -Wno-stringop-overflow -Wno-array-bounds -Wno-alloc-size-larger-than -Wimplicit-fallthrough=5 -Werror=date-time -Werror=incompatible-pointer-types -Werror=designated-init -Wenum-conversion -Wno-unused-but-set-variable -Wno-unused-const-variable -Wno-restrict -Wno-packed-not-aligned -Wno-format-overflow -Wno-format-truncation -Wno-stringop-truncation -Wno-missing-field-initializers -Wno-type-limits -Wno-shift-negative-value -Wno-maybe-uninitialized -Wno-sign-compare -g -fsanitize=kernel-address -fasan-shadow-offset=0xdffffc0000000000 --param asan-globals=1 --param asan-instrumentation-with-call-threshold=10000 --param asan-instrument-allocas=1 --param asan-stack=1 --param asan-kernel-mem-intrinsic-prefix=1 -DMODULE -DKBUILD_BASENAME="super" -DKBUILD_MODNAME="ext4" -D__KBUILD_MODNAME=kmod_ext4 -c -o fs/ext4/super.o fs/ext4/super.c -DCE_DEBUGINFO_PATH=fs/ext4/ext4.ko -DCE_SYMVERS_PATH=Module.symvers -DCE_OUTPUT_FILE=/tmp/bsc888_fs_ext4_super.c -DCE_EXTRACT_FUNCTIONS=ext4_remount -DCE_KEEP_INCLUDES -DCE_RENAME_SYMBOLS -DCE_DUMP_PASSES
```

(You would need EXT4_FS enabled as module and a kernel tree compiled for this to work)

The error shown is
```c
In file included from fs/ext4/super.c:1:                                                                                            
In file included from fs/ext4/super.c:20:                                                                                           
In file included from /home/mpdesouza/git/linux/include/linux/module.h:12:                                      
/home/mpdesouza/git/linux/include/linux/list.h:1063:2: error: attribute 'error' is already applied with different arguments         
 1063 |         WRITE_ONCE(n->pprev, &prev->next);                                                                                  
      |         ^                                                                                                                   
/home/mpdesouza/git/linux/include/asm-generic/rwonce.h:60:2: note: expanded from macro 'WRITE_ONCE'                                 
   60 |         compiletime_assert_rwonce_type(x);                              \                                                   
      |         ^                                                                                                                   
/home/mpdesouza/git/linux/include/asm-generic/rwonce.h:36:2: note: expanded from macro 'compiletime_assert_rwonce_type'             
   36 |         compiletime_assert(__native_word(t) || sizeof(t) == sizeof(long long),  \                                           
      |         ^                                                                                                                   
/home/mpdesouza/git/linux/include/linux/compiler_types.h:442:2: note: expanded from macro 'compiletime_assert'
  442 |         _compiletime_assert(condition, msg, __compiletime_assert_, __COUNTER__)                                             
      |         ^                                                                                                                   
/home/mpdesouza/git/linux/include/linux/compiler_types.h:430:2: note: expanded from macro '_compiletime_assert'
  430 |         __compiletime_assert(condition, msg, prefix, suffix)                                                                
      |         ^                                                                                                                   
/home/mpdesouza/git/linux/include/linux/compiler_types.h:421:4: note: expanded from macro '__compiletime_assert'                    
  421 |                         __compiletime_error(msg);
...
/home/mpdesouza/git/linux/include/linux/compiler_types.h:421:4: note: expanded from macro '__compiletime_assert'
  421 |                         __compiletime_error(msg);                       \
      |                         ^
/home/mpdesouza/git/linux/include/linux/compiler_attributes.h:151:56: note: expanded from macro '__compiletime_error'
  151 | # define __compiletime_error(msg)       __attribute__((__error__(msg)))
      |                                                        ^

Error on pass: BuildASTPass
```

And the same "error is already applied" is repeated for other headers too. As the problem happens when building the AST clang-extract doesn't generate any dump. Please let me know if you need more info.

I had to remove a couple of arguments used by gcc, and now it returns an error about structs:

```
clang -Wp,-MMD,fs/ext4/.super.o.d -nostdinc -I./arch/x86/include -I./arch/x86/include/generated -I./include -I./arch/x86/include/uapi -I./arch/x86/include/generated/uapi -I./include/uapi -I./include/generated/uapi -include ./include/linux/compiler-version.h -include ./include/linux/kconfig.h -include ./include/linux/compiler_types.h -D__KERNEL__ -fmacro-prefix-map=./= -Werror -std=gnu11 -fshort-wchar -funsigned-char -fno-common -fno-PIE -fno-strict-aliasing -mno-sse -mno-mmx -mno-sse2 -mno-3dnow -mno-avx -fcf-protection=branch -fno-jump-tables -m64 -falign-loops=1 -mno-80387 -mno-fp-ret-in-387 -mskip-rax-setup -mtune=generic -mno-red-zone -mcmodel=kernel -Wno-sign-compare -fno-asynchronous-unwind-tables -mindirect-branch-cs-prefix -mfunction-return=thunk-extern -fno-jump-tables -fpatchable-function-entry=16,16 -fno-delete-null-pointer-checks -O2 -fstack-protector-strong -ftrivial-auto-var-init=zero -fno-stack-clash-protection -pg -mfentry -DCC_USING_FENTRY -falign-functions=16 -fstrict-flex-arrays=3 -fno-strict-overflow -fno-stack-check -Wall -Wundef -Werror=implicit-function-declaration -Werror=implicit-int -Werror=return-type -Werror=strict-prototypes -Wno-format-security -Wno-trigraphs -Wno-frame-address -Wno-address-of-packed-member -Wmissing-declarations -Wmissing-prototypes -Wframe-larger-than=2048 -Wno-main -Wno-unused-but-set-variable -Wno-unused-const-variable -Wvla -Wno-pointer-sign -Wcast-function-type -Wno-array-bounds -Werror=date-time -Werror=incompatible-pointer-types -Wenum-conversion -Wno-unused-but-set-variable -Wno-unused-const-variable -Wno-format-overflow -Wno-format-truncation -Wno-missing-field-initializers -Wno-type-limits -Wno-shift-negative-value -Wno-sign-compare -g -fsanitize=kernel-address -DMODULE -DKBUILD_BASENAME="super" -DKBUILD_MODNAME="ext4" -D__KBUILD_MODNAME=kmod_ext4 -c -o fs/ext4/super.o fs/ext4/super.c -DCE_DEBUGINFO_PATH=fs/ext4/ext4.ko -DCE_SYMVERS_PATH=Module.symvers -DCE_OUTPUT_FILE=/tmp/bsc888_fs_ext4_super.c -DCE_EXTRACT_FUNCTIONS=ext4_remount -DCE_KEEP_INCLUDES -DCE_RENAME_SYMBOLS -DCE_DUMP_PASSES
```

The error:
```
In file included from fs/ext4/super.c:52:
In file included from fs/ext4/ext4.h:24:
./include/linux/jbd2.h:1790:21: error: field 'shash' with variable sized type 'struct shash_desc' not at the end of a struct or class is a GNU extension [-Werror,-Wgnu-variable-sized-type-not-at-end]
 1790 |                 struct shash_desc shash;
      |                                   ^
In file included from fs/ext4/super.c:52:
fs/ext4/ext4.h:2467:21: error: field 'shash' with variable sized type 'struct shash_desc' not at the end of a struct or class is a GNU extension [-Werror,-Wgnu-variable-sized-type-not-at-end]
 2467 |                 struct shash_desc shash;
      |                                   ^
In file included from fs/ext4/super.c:57:
In file included from fs/ext4/mballoc.h:17:
In file included from ./include/linux/swap.h:9:
In file included from ./include/linux/memcontrol.h:13:
In file included from ./include/linux/cgroup.h:28:
./include/linux/cgroup-defs.h:574:16: error: field 'cgrp' with variable sized type 'struct cgroup' not at the end of a struct or class is a GNU extension [-Werror,-Wgnu-variable-sized-type-not-at-end]
  574 |         struct cgroup cgrp;
      |                       ^
3 errors generated.
```

But if I add -Wno-gnu-variable-sized-type-not-at-end, it solves the problem, as we can compile using clang.

--- a/libcextract/Passes.cpp
+++ b/libcextract/Passes.cpp
@@ -79,7 +79,7 @@ static bool Build_ASTUnit(PassManager::Context *ctx, IntrusiveRefCntPtr<vfs::Fil
   PCHContainerOps = std::make_shared<PCHContainerOperations>();
 
   auto AU = ASTUnit::LoadFromCompilerInvocation(
-      CInvok, PCHContainerOps, Diags, FileMgr, false, CaptureDiagsKind::None, 1,
+      CInvok, PCHContainerOps, Diags, FileMgr, false, CaptureDiagsKind::None, 0,
       TU_Complete, false, false, false);
 
   if (AU == nullptr) {

--- a/testsuite/includes/pch-1.c
+++ b/testsuite/includes/pch-1.c
@@ -1,0 +1,13 @@
+/* { dg-options "-DCE_EXTRACT_FUNCTIONS=main -DCE_NO_EXTERNALIZATION" }*/
+
+#include "pch-1.h"
+
+DEFINE_FUNCTION;
+
+int main()
+{
+  function_0();
+  return 0;
+}
+
+/* { dg-final { scan-tree-dump "function_0" } } */

--- a/testsuite/includes/pch-1.h
+++ b/testsuite/includes/pch-1.h
@@ -1,0 +1,12 @@
+#define __DEFINE_FUNCTION(i) \
+  int function_ ## i (void) { \
+    return 0;\
+  }
+
+#define _DEFINE_FUNCTION(i) \
+  __DEFINE_FUNCTION(i)
+
+#define DEFINE_FUNCTION \
+  _DEFINE_FUNCTION(__COUNTER__)
+
+DEFINE_FUNCTION;

